src/common/configIntelliSenseProvider.ts

```diff
@@ -0,0 +1,336 @@
+        }
+
+        // Disable following two group of completion items temporarily because they will be duplicate with built-in JSON completion items
+        // Tracking issue: https://github.com/Microsoft/vscode/issues/45864
+
+        // if (location.path[0] === "moduleContent" && location.path[1] === "$edgeAgent"
+        //     && location.path[2] === "properties.desired" && location.path[3] === "modules"
+        //     && location.path[5] === "status") {
+        //     return this.getCompletionItems(Constants.moduleStatuses, document, position);
+        // }
+
+        // if (location.path[0] === "moduleContent" && location.path[1] === "$edgeAgent"
+        //     && location.path[2] === "properties.desired" && location.path[3] === "modules"
+        //     && location.path[5] === "restartPolicy") {
+        //     return this.getCompletionItems(Constants.moduleRestartPolicies, document, position);
+        // }
+
+        if (this.locationMatch(location, Constants.imgDeploymentManifestJsonPath)) {
+            const images: string[] = await this.getSlnImgPlaceholders(document.uri);
+            return this.getCompletionItems(images, document, position, location);
+        }
+
+        if (this.locationMatch(location, Constants.routeDeploymentManifestJsonPath)) {
+            const json = parser.parse(document.getText());
+            const modules: any = ((json.moduleContent.$edgeAgent || {})["properties.desired"] || {}).modules || {};
+            const moduleIds: string[] = Object.keys(modules);
+
+            const routeCompletionItem: vscode.CompletionItem = new vscode.CompletionItem(Constants.routeSnippetLabel);
+            routeCompletionItem.filterText = `\"${Constants.routeSnippetLabel}\"`;
+            routeCompletionItem.kind = vscode.CompletionItemKind.Snippet;
+            routeCompletionItem.detail = Constants.routeSnippetDetail;
+            routeCompletionItem.range = document.getWordRangeAtPosition(position);
+            routeCompletionItem.insertText = new vscode.SnippetString(this.getRouteSnippetString(moduleIds));
+            return [routeCompletionItem];
+        }
+    }
+
+    public async provideHover(document: vscode.TextDocument, position: vscode.Position): Promise<vscode.Hover> {
+        const location: parser.Location = parser.getLocation(document.getText(), document.offsetAt(position));
+
+        if (this.locationMatch(location, Constants.imgDeploymentManifestJsonPath)) {
+            const moduleToImageMap: Map<string, string> = new Map();
+            const imageToDockerfileMap: Map<string, string> = new Map();
+
+            try {
+                await Utility.setSlnModulesMap(path.dirname(document.uri.fsPath), moduleToImageMap, imageToDockerfileMap);
+
+                const node: parser.Node = location.previousNode;
+                const imageUrl: string = node.value;
+                const imageUrlUnwrapped: string = imageUrl.slice(2, -1); // remove the wrapping "${" and "}"
```

Comment:
edge case. If the imageUrl is not start with ${ and end with }, And after slicing imageUrl.slice(2, -1), the value could be got from the map.

Reply:
Fixed. Thanks.